text/0000-gradual-struct-init.md

```diff
@@ -0,0 +1,739 @@
+
+```rust
+struct Foo<T> {
+    bar: usize,
+    qux: T,
+}
+```
+
+Suppose also that you have a function that consumes a `Foo<T>` (2):
+
+```rust
+fn consume<T>(_foo: Foo<T>) {}
+```
+
+Currently, if you want to make a `Foo` then you have to write (3):
+
+```rust
+let foo = Foo {
+    bar: 42usize,
+    qux: 24u8,
+};
+```
+
+**_We propose_** that you should be able to also write (4):
+
+```rust
+let foo;
+foo.bar = 42usize;
+foo.qux = 24u8;
+
+consume(foo); // OK!
+```
+
+### Immutable and mutable bindings
+
+Note that in snippet (4), you are *not* mutating `foo`, `foo.bar`, or `foo.qux`.
+These fields are only being initialized. If you however assign to `foo.bar`
+twice as in (5):
+
+```rust
+let foo: Foo<u8>;
+
+foo.bar = 42usize;
+foo.qux = 24u8;
+
+// Error! Second time you assign to `foo.bar` but `foo` is not a mutable binding!
+foo.bar = 43;
+
+drop(foo);
+```
+
+then the compiler will reject the program as ill-formed because
+`foo` was not declared as a *mutable* binding with `mut foo`.
+To remedy this, **_you may write_** (6):
+
+```rust
+let mut foo: Foo<u8>;
+
+foo.bar = 42usize;
+foo.qux = 24u8;
+
+foo.bar = 43; // OK!
+
+drop(foo);
+```
+
+### Definitive initialization and usage
+
+Note that in both (4) and (6), you can only `drop(foo)`, `consume(foo)`,
+or take references to `foo` because you have fully initialized `foo`.
+Suppose instead that you wrote (7):
+
+```rust
+let mut foo;
+
+foo.qux = 42;
+
+consume(foo); // Error! `foo.bar` is not initialized.
+```
+
+The snippet in (7) would be *rejected* by the compiler because
+otherwise `foo.bar` would be uninitialized and thus you would not
+have a valid `Foo`, thus causing the type system to be unsound and
+the program to exhibit undefined behaviour.
+
+In our proposal, it is also not legal to take a reference to `foo`, as with
+`&mut foo` or `&foo` while parts of it isn't initialized. You also cannot
+copy `foo` before it is fully initialized. Neither may you take references to,
+move, or copy the parts of `foo` (in this case `foo.bar`) that are not yet
+initialized. This extends to conditional initialization. **_The compiler will
+allow you to write (8)_**:
+
+```rust
+let foo;
+
+foo.bar = 42;
+
+if random_bool() {
+    foo.qux = 1;
+} else {
+    foo.qux = 2;
+}
+
+// Nota bene:
+// In this case foo.qux = if random_bool() { 1 } else { 2 };
+// is preferable as a matter of style.
+
+consume(foo); // OK!
+```
+
+The snippet in (8) is allowed because when `consume(foo)` is reached, `foo.qux`
+is provably initialized no matter what branch is taken in the conditional.
+However, the following snippet would be rejected because `foo.qux` may be
+uninitialized in the `else` branch (9):
+
+```rust
+let foo;
+
+foo.bar = 42;
+
+if random_bool() {
+    foo.qux = 1;
+} else {
+    // nothing here.
+}
+
+consume(foo); // ERROR! `foo.qux` not initialized in `else { .. }`.
+```
+
+The general condition here is that all fields must be *definitely initialized*.
```

Comment:
Similarly I assume that

```rust
let foo;

foo.bar = 42;

if random_bool() {
    foo.qux = 1;
} else {
    // nothing here.
}

foo.qux = 2; // ERROR! `foo.qux` initialized in `if { .. }`.

consume(foo);
```

will error because `foo.qux` may be initialized in the `if` branch (and `foo` is not mutable).

Reply:
Yup, that's correct. Maybe I should clarify somehow / include this example...